sdks/python/apache_beam/utils/pipeline_options.py

```diff
@@ -20,6 +20,69 @@
 import argparse
 
 from apache_beam.transforms.display import HasDisplayData
+from apache_beam.utils.value_provider import StaticValueProvider
+from apache_beam.utils.value_provider import RuntimeValueProvider
+from apache_beam.utils.value_provider import ValueProvider
+
+
+def static_value_provider_of(value_type):
+  """"Helper function to plug a ValueProvider into argparse.
+
+  Args:
+    value_type: the type of the value. Since the type param of argparse's
+                add_argument will always be ValueProvider, we need to
+                preserve the type of the actual value.
+  Returns:
+    A partially constructed StaticValueProvider in the form of a function.
+
+  """
+  def _f(value):
+    _f.func_name = value_type.__name__
+    return StaticValueProvider(value_type, value)
+  return _f
+
+
+class ValueProviderArgumentParser(argparse.ArgumentParser):
+  """This class provides an API to add options of ValueProvider type.
+
+  It preserves the functionalities of the parent ArgumentParser.
+  A template user willing to define parameterizable options will
+  only need to define a subclass of PipelineOptions in this manner:
+
+    class TemplateUserOptions(PipelineOptions):
+
+      @classmethod
+      def _add_argparse_args(cls, parser):
+          parser.add_value_provider_argument('--abc', default='start')
+          parser.add_value_provider_argument('--xyz', default='end')
+  """
+  def add_value_provider_argument(self, *args, **kwargs):
+    # extract the option name.
+    # TODO(mariapython): handle multiple positional args like ('--quux',)
+    option_name = args[0].replace('-', '')
+
+    # reassign the type to make room for StaticValueProvider
+    value_type = kwargs.get('type') or str
+
+    # use StaticValueProvider as the type of the argument
+    kwargs['type'] = static_value_provider_of(value_type)
+
+    # reassign default to value_default to make room for using
+    # RuntimeValueProvider as the default for add_argument
+    default_value = kwargs.get('default')
+
+    # use RuntimeValueProvider() as the default
+    kwargs['default'] = RuntimeValueProvider(
+        pipeline_options_subclass=(self.pipeline_options_subclass
+                                   or PipelineOptions),
+        option_name=option_name,
+        value_type=value_type,
+        default_value=default_value,
+    )
+    kwargs['nargs'] = '?'  # make positional arguments optionally templated
+
+    # we still want add_argument to do most of the work
```

Comment:
you can remove some of these comments if the next line of code is capturing the same thing as the comment.